packages/flutter/lib/src/widgets/focus_traversal.dart

```diff
@@ -695,136 +700,317 @@ class _SortData {
+      final List<_ReadingOrderSortData> inBandOfTop = inBand(topmost, candidates).toList();
       inBandOfTop.sort(compareBeginningSide);
       if (inBandOfTop.isNotEmpty) {
         return inBandOfTop.first;
       }
       return topmost;
     }
 
-    final List<_SortData> data = <_SortData>[
-      for (final FocusNode node in nodes) _SortData(node),
+    final List<_ReadingOrderSortData> data = <_ReadingOrderSortData>[
+      for (final FocusNode node in nodes) _ReadingOrderSortData(node),
     ];
 
     // Pick the initial widget as the one that is leftmost in the band of the
     // topmost, or the topmost, if there are no others in its band.
-    final List<_SortData> sortedList = <_SortData>[];
-    final List<_SortData> unplaced = data.toList();
-    _SortData current = pickFirst(unplaced);
+    final List<_ReadingOrderSortData> sortedList = <_ReadingOrderSortData>[];
+    final List<_ReadingOrderSortData> unplaced = data.toList();
+    _ReadingOrderSortData current = pickFirst(unplaced);
     sortedList.add(current);
     unplaced.remove(current);
 
     while (unplaced.isNotEmpty) {
-      final _SortData next = pickFirst(unplaced);
+      final _ReadingOrderSortData next = pickFirst(unplaced);
       current = next;
       sortedList.add(current);
       unplaced.remove(current);
     }
-    return sortedList.map((_SortData item) => item.node);
+    return sortedList.map((_ReadingOrderSortData item) => item.node);
   }
+}
 
-  // Moves the focus forward or backward in reading order, depending on the
-  // value of the forward argument.
-  bool _move(FocusNode currentNode, {@required bool forward}) {
-    final FocusScopeNode nearestScope = currentNode.nearestScope;
-    invalidateScopeData(nearestScope);
-    final FocusNode focusedChild = nearestScope.focusedChild;
-    if (focusedChild == null) {
-      final FocusNode firstFocus = findFirstFocus(currentNode);
-      if (firstFocus != null) {
-        _focusAndEnsureVisible(
-          firstFocus,
-          alignmentPolicy: forward
-              ? ScrollPositionAlignmentPolicy.keepVisibleAtEnd
-              : ScrollPositionAlignmentPolicy.keepVisibleAtStart,
-        );
-        return true;
+/// Base class for all sort keys for [OrderedFocusTraversalPolicy] traversal.
+///
+/// Only orders of the same type are comparable. If a set of widgets in the same
+/// [FocusScope] contains keys that are not comparable with each other, it will
+/// assert, since the ordering is undefined.
+///
+/// See also:
+///
+///  * [NumericFocusOrder], for a focus order that describes its order with a
+///    `double`.
+abstract class FocusOrder extends Diagnosticable implements Comparable<FocusOrder> {
+  /// Abstract const constructor. This constructor enables subclasses to provide
+  /// const constructors so that they can be used in const expressions.
+  const FocusOrder();
+
+  @override
+  int compareTo(FocusOrder other) {
```

Comment:
Looks like subclasses should override doCompare and never override compareTo? That should be documented somewhere and we could also mark compareTo as `nonVirtual`?